### ISSUE TITLE
About apply Qlora for DPO training

### ISSUE THREAD

**Issue by Iambestfeed** [eric-mitchell/direct-preference-optimization#1]
Hello,
I don't think this is an issue but this is just my question about whether Qlora can be applied to DPO and if so what are the constraints on batch_size or model size? (I tested the Reward model and it seems that the model size, batch_size affect quite a lot the convergence ability).
It would be great if you could suggest me what to tweak to be able to incorporate Qlora into DPO based training.
Thanks for this repo.

**Comment by eric-mitchell** [comment]
If you're training on a single GPU using the `BasicTrainer`, I don't expect there would be any issue doing DPO with Qlora, though we haven't tested it. There will probably be interactions between quantization and FSDP (see [this HuggingFace issue](https://github.com/pytorch/pytorch/issues/91165)).

The constraints on batch size and model size will depend on the model you're fine-tuning.

If you want to integrate Qlora, I'd start [here](https://github.com/eric-mitchell/direct-preference-optimization/blob/main/train.py#L78), which is where the model is created. Integrating [HF PEFT](https://github.com/huggingface/peft), at least, should be straightforward here.

Feel free to re-open if you've got more questions!

**Comment by Iambestfeed** [comment]
@eric-mitchell 
Hello, I tried apply Qlora and it seems I succeeded in doing this (I trained sft on model Bloom3b with Kaggle P100 only).
However I have problem with loss (see picture below)
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/95571916/dd60e8cb-e7b7-4c14-bb8b-e4d0ec1f4a4c)
Don't care about the name because I mistyped pythia28 instead of bloom3b (but it seems the loss train is too unstable). My question is is the problem is because I keep the batch size too small (only 4) so ​​the training becomes inefficient.

**Comment by eric-mitchell** [comment]
These learning curves actually look okay to me- there are 150 warmup steps for the learning rate by default (measured in batches, whereas your x axis here is measured in examples). So I'd at least wait until the warmup is done and see if the loss starts really decreasing then.

**Comment by Iambestfeed** [comment]
@eric-mitchell 
I have set batch_size to 12 and default warnup step to 150 but it seems like things are getting worse isn't it?
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/95571916/1ae3b6e0-a6fd-45a6-ace6-fa6fddb1fad9)
Or have things not gone through the warnup step yet and am I just worrying too much?

**Comment by eric-mitchell** [comment]
Are you sure you're running DPO, and not SFT? The scale of the loss looks larger than I'd expect for DPO. Can you share the curves under the `rewards_train` pane? Can you share the exact launch command logged in the wandb info page for the run?

There shouldn't be any problem with larger models; I just picked pythia 2.8B fairly arbitrary as an example. I've trained llama 7B and 13B on Anthropic HH + Stanford Human Preferences with this codebase without any issues (learning curves below).

<img width="681" alt="Screenshot 2023-06-24 at 9 43 22 PM" src="https://github.com/eric-mitchell/direct-preference-optimization/assets/56408839/3f2dd329-71bb-45ac-b2fc-12030fc78a03">

**Comment by Iambestfeed** [comment]
> Are you sure you're running DPO, and not SFT? The scale of the loss looks larger than I'd expect for DPO. Can you share the curves under the `rewards_train` pane? Can you share the exact launch command logged in the wandb info page for the run?
> 
> There shouldn't be any problem with larger models; I just picked pythia 2.8B fairly arbitrary as an example. I've trained llama 7B and 13B on Anthropic HH + Stanford Human Preferences with this codebase without any issues (learning curves below).
> 
> <img alt="Screenshot 2023-06-24 at 9 43 22 PM" width="681" src="https://user-images.githubusercontent.com/56408839/248529271-3f2dd329-71bb-45ac-b2fc-12030fc78a03.png">

I'm training SFT and it seems the only problem here is that the batch_size is too small which leads me to be a bit nervous about things.
This may be a bit of an exaggerated request but can you share me the wandb results of the model LLama 7B of both SFT phase and DPO phase?

**Comment by eric-mitchell** [comment]
I've added the wandb for the 7b and 13b llama runs [here](https://wandb.ai/eric_anthony_mitchell/dpo-demos). Keep in mind these were from a slightly earlier version of the codebase, so some logging metrics are named differently (and the x axis is microbatches, not examples).

Can you please share the exact launch command logged in the wandb info page for the run?

**Comment by eric-mitchell** [comment]
@Iambestfeed Just checking in- are you still running into issues with qlora?

**Comment by Iambestfeed** [comment]
> @Iambestfeed Just checking in- are you still running into issues with qlora?
Oh sorry, I got it running but it seems I forgot to check the issues I opened. Thank you so much for your support
